src/uint/boxed/shl.rs

```diff
@@ -1,108 +1,127 @@
 //! [`BoxedUint`] bitwise left shift operations.
 
-use crate::{BoxedUint, CtChoice, Limb, Word};
+use crate::{BoxedUint, Limb};
 use core::ops::{Shl, ShlAssign};
 use subtle::{Choice, ConstantTimeLess};
 
 impl BoxedUint {
     /// Computes `self << shift`.
-    /// Returns zero if `shift >= Self::BITS`.
-    pub fn shl(&self, shift: u32) -> Self {
+    /// Returns `None` if `shift >= Self::BITS`.
+    pub fn shl(&self, shift: u32) -> (Self, Choice) {
+        // `floor(log2(bits_precision - 1))` is the number of bits in the representation of `shift`
+        // (which lies in range `0 <= shift < bits_precision`).
+        let shift_bits = u32::BITS - (self.bits_precision() - 1).leading_zeros();
         let overflow = !shift.ct_lt(&self.bits_precision());
         let shift = shift % self.bits_precision();
-        let log2_bits = u32::BITS - self.bits_precision().leading_zeros();
         let mut result = self.clone();
+        let mut temp = self.clone();
 
-        for i in 0..log2_bits {
+        for i in 0..shift_bits {
             let bit = Choice::from(((shift >> i) & 1) as u8);
-            result = Self::conditional_select(&result, &result.shl_vartime(1 << i), bit);
+            temp.set_to_zero();
+            // Will not overflow by construction
+            result
+                .shl_vartime_into(&mut temp, 1 << i)
+                .expect("shift within range");
+            result.conditional_assign(&temp, bit);
         }
 
-        Self::conditional_select(
-            &result,
-            &Self::zero_with_precision(self.bits_precision()),
+        (
+            Self::conditional_select(
+                &result,
+                &Self::zero_with_precision(self.bits_precision()),
+                overflow,
+            ),
```

Comment:
While you're here, it seems like this could be replaced with a loop over the limbs which does `conditional_assign(&Limb::ZERO, overflow)` and thus avoids an allocation

Reply:
Added `BoxedUint::conditional_set_to_zero()` for that